parse error recovery could be better with missing commas after struct fields

Test case: [(playground link)](https://play.rust-lang.org/?gist=67235946f5140d25364ffacba39cd3b0&version=nightly&mode=debug)
```rust
struct S {
    foo: u32 // <- missing comma
    bar: u32
}

fn main() {
    let s = S { foo: 5, bar: 6 };
}
```

Actual result (nightly):
```
error: expected `,`, or `}`, found `bar`
 --> src/main.rs:3:5
  |
3 |     bar: u32
  |     ^^^
  |
  = help: struct fields should be separated by commas

error[E0560]: struct `S` has no field named `foo`
 --> src/main.rs:7:17
  |
7 |     let s = S { foo: 5, bar: 6 };
  |                 ^^^ `S` does not have this field

error[E0560]: struct `S` has no field named `bar`
 --> src/main.rs:7:25
  |
7 |     let s = S { foo: 5, bar: 6 };
  |                         ^^^ `S` does not have this field
```

Expected result:

Ideally the parser would guess that there was meant to be a comma at the end of the line.  Then:

- The "expected comma" error could highlight the line missing a comma (and suggest a fix-it), rather than highlighting the first token in the next line.
- After producing an error, it could recover by pretending the comma was there, adding the `foo` and `bar` fields to the struct to avoid the latter two errors.

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -5797,9 +5797,18 @@ impl<'a> Parser<'a> {
                     return Err(err);
                 }
             }
-            _ => return Err(self.span_fatal_help(self.span,
-                    &format!("expected `,`, or `}}`, found `{}`", self.this_token_to_string()),
-                    "struct fields should be separated by commas")),
+            _ => {
+                let sp = self.sess.codemap().next_point(self.prev_span);
+                let mut err = self.struct_span_err(sp, &format!("expected `,`, or `}}`, found `{}`",
+                                                                self.this_token_to_string()));
+                if self.token.is_ident() {
+                    // This is likely another field; emit the diagnostic and keep going
+                    err.span_suggestion(sp, "try adding a comma", ",".into());
+                    err.emit();
+                } else {
+                    return Err(err)
+                }
+            }
         }
         Ok(a_var)
     }

--- a/src/test/ui/struct-missing-comma.rs
+++ b/src/test/ui/struct-missing-comma.rs
@@ -1,0 +1,23 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// compile-flags: -Z parse-only
+
+// Issue #50636
+
+struct S {
+    foo: u32 //~ expected `,`, or `}`, found `bar`
+    //     ~^ HELP try adding a comma: ','
+    bar: u32
+}
+
+fn main() {
+    let s = S { foo: 5, bar: 6 };
+}

--- a/src/test/ui/struct-missing-comma.stderr
+++ b/src/test/ui/struct-missing-comma.stderr
@@ -1,0 +1,8 @@
+error: expected `,`, or `}`, found `bar`
+  --> $DIR/struct-missing-comma.rs:16:13
+   |
+LL |     foo: u32 //~ expected `,`, or `}`, found `bar`
+   |             ^ help: try adding a comma: `,`
+
+error: aborting due to previous error
+

